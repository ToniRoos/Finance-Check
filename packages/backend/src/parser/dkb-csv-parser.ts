import { AccountData } from "common";
import _ from "lodash";

const regexForDate = /Buchungstag|Belegdatum/;
const regexForAmount = /Betrag \(EUR\)/;
const regexForClient = /Auftraggeber|Beschreibung/;
const regexForCreditor = /Gl.ubiger-ID/;
const regexForBankAccountNumber = /Kontonummer/;
const regexForReason = /Verwendungszweck/;

export function parseCsv(content: string) {

    console.log("Start import");

    if (_.isEmpty(content)) {
        throw new Error('CSV content is empty')
    }

    const contentCleaned = content.replaceAll('"', '')
    const lines = contentCleaned.split('\n')

    const data: any = [];
    let saldo = 0;
    const seperator = ';'

    let colForDate = 0;
    let colForAmount = 0;
    let colForClient = 0;
    let colForCreditor = 0;
    let colForBankAccountNumber = 0;
    let colForReason = 0;

    const metaEndIndex = _.findIndex(lines, line => regexForDate.test(line))
    if (metaEndIndex < 0) {
        throw new Error('Invalid csv file. Meta end not found')
    }

    const metaLines = lines.splice(0, metaEndIndex + 1)
    const [__, bankAccountNumber] = metaLines[0].split(seperator)
    if (_.isEmpty(bankAccountNumber)) {
        throw new Error('Invalid csv file. Bank account number not found')
    }

    const [___, saldoString] = metaLines[4].split(seperator);
    const [saldoStringValue] = saldoString.split(' ')
    if (Number(saldoStringValue)) {
        saldo = parseFloat(saldoStringValue);
    } else {
        saldo = parseFloat(saldoStringValue.replace('.', '').replace(',', '.'));
    }

    const headers = metaLines[metaEndIndex].split(seperator)
    for (let index = 0; index < headers.length; index++) {

        const header = headers[index];
        if (header.match(regexForDate)) {
            colForDate = index;
        }
        if (header.match(regexForAmount)) {
            colForAmount = index;
        }
        if (header.match(regexForClient)) {
            colForClient = index;
        }
        if (header.match(regexForCreditor)) {
            colForCreditor = index;
        }
        if (header.match(regexForBankAccountNumber)) {
            colForBankAccountNumber = index;
        }
        if (header.match(regexForReason)) {
            colForReason = index;
        }
    }

    for (const line of lines) {
        const cols = line.split(';')

        data.push({
            bookingDate: parseDate(cols[colForDate]),
            client: cols[colForClient],
            reason: cols[colForReason],
            creditor: cols[colForCreditor],
            bankAccountNumber: cols[colForBankAccountNumber],
            amount: parseAmount(cols[colForAmount])
        });
    }

    console.log("Finished import");
    const result: AccountData = { data, bankAccountNumber, saldo }
    return result
}

function parseDate(dateString: string) {
    const [day, month, year] = dateString.split('.')
    const parsedDate = new Date(`${year}-${month}-${day}`);
    return parsedDate
}

function parseAmount(amounString: string) {
    const amountParsed = parseFloat(amounString.replace('.', '').replace(',', '.').replace(' ', ''));
    return amountParsed;
}