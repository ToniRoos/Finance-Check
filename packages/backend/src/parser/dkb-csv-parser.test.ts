import { readFile } from "fs-extra"
import { resolve } from "path"
import { parseCsv } from "./dkb-csv-parser"

describe('dkb csv parser', () => {

    const testDataDir = resolve(__dirname, '__fixtures__')

    it('should parse csv correctly [ csv1 ]', async () => {

        // assign
        const content = await readFile(resolve(testDataDir, 'csv1.csv'))

        // act
        const parsedContent = parseCsv(content.toString())

        //assert
        expect(parsedContent).toMatchSnapshot()
    })

    it('should parse csv correctly [ csv2 ]', async () => {

        // assign
        const content = await readFile(resolve(testDataDir, 'csv2.csv'))

        // act
        const parsedContent = parseCsv(content.toString())

        //assert
        expect(parsedContent).toMatchSnapshot()
    })
})