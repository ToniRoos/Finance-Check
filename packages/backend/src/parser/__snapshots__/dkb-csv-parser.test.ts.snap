// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dkb csv parser should parse csv correctly [ csv1 ] 1`] = `
{
  "bankAccountNumber": "DE1234567889 / Girokonto",
  "data": [
    {
      "amount": -22.22,
      "bankAccountNumber": "DE112233445566",
      "bookingDate": 2020-11-04T00:00:00.000Z,
      "client": "Foo GmbH + Co. OHG",
      "creditor": "DE9342347962",
      "reason": "Kd-Nr.: 31111111, Rg-Nr.: 123456/6, Ihre Tarifrechnung",
    },
    {
      "amount": -36.58,
      "bankAccountNumber": "DE7460023423434231234",
      "bookingDate": 2020-11-04T00:00:00.000Z,
      "client": "Bar GMBH / Bar GMBH",
      "creditor": "DE39DB2312300002177",
      "reason": "u7d67z7d3 11/2020, Max Musterman, 67633762736-001 NR 1016556",
    },
  ],
  "saldo": 2958.32,
}
`;

exports[`dkb csv parser should parse csv correctly [ csv2 ] 1`] = `
{
  "bankAccountNumber": "DE734234324131 / Girokonto",
  "data": [
    {
      "amount": 9999.99,
      "bankAccountNumber": "DE3423478923748924",
      "bookingDate": 2022-10-27T00:00:00.000Z,
      "client": "Foo GmbH DE",
      "creditor": "",
      "reason": "Lohn adsad939d/372837",
    },
    {
      "amount": -55.55,
      "bankAccountNumber": "DE982034923840234",
      "bookingDate": 2022-10-24T00:00:00.000Z,
      "client": "SHELL 32432432423",
      "creditor": "",
      "reason": "2022-10-23 Debit",
    },
  ],
  "saldo": 3444.08,
}
`;
